lib/Sema/SemaType.cpp

```diff
@@ -3725,68 +3741,54 @@ QualType Sema::MakeCheckedArrayType(QualType T, bool Diagnose,
+  if (Ty.isNull() || Bounds == nullptr)
     return ResultType;
 
-  if (const BoundsExpr *Bounds = TargetDecl->getBoundsExpr()) {
-    switch (Bounds->getKind()) {
-      case BoundsExpr::Kind::InteropTypeAnnotation: {
-        const InteropTypeBoundsAnnotation *Annot =
-          dyn_cast<InteropTypeBoundsAnnotation>(Bounds);
-        assert(Annot && "unexpected dyn_cast failure");
-        if (Annot != nullptr)
-          ResultType = Annot->getType();
-        break;
-      }
-      case BoundsExpr::Kind::ByteCount:
-      case BoundsExpr::Kind::ElementCount:
-      case BoundsExpr::Kind::Range: {
-        QualType Ty;
-        // The types for parameter variables that have array types are adjusted
-        // to be pointer type.  We'll work with the original array type instead.
-        // For multi-dimensional array types, the nested array types need to
-        // become checked array types too.
-        if (const ParmVarDecl *ParmVar = dyn_cast<ParmVarDecl>(TargetDecl))
-          Ty = ParmVar->getOriginalType();
-        else if (const FunctionDecl *FD = dyn_cast<FunctionDecl>(TargetDecl))
-          Ty = FD->getReturnType();
-        else
-          Ty = Decl->getType();
+  // Parameter types that are array types are adusted to be
+  // pointer types.  We'll work with the original type instead.
+  // For multi-dimensional array types, the nested array types need
+  // to become checked array types too.
+  if (isParam) {
+    if (const DecayedType *Decayed = dyn_cast<DecayedType>(Ty)) {
+      Ty = Decayed->getOriginalType();
+    }
+  }
 
-        if (const PointerType *PtrType = Ty->getAs<PointerType>()) {
-          if (PtrType->isUnchecked()) {
-            ResultType = Context.getPointerType(PtrType->getPointeeType(),
-                                                CheckedPointerKind::Array);
-            ResultType.setLocalFastQualifiers(Ty.getCVRQualifiers());
-          }
-        }
-        else if (Ty->isConstantArrayType() || Ty->isIncompleteArrayType()) {
-          ResultType = MakeCheckedArrayType(Ty);
+  switch (Bounds->getKind()) {
+    case BoundsExpr::Kind::InteropTypeAnnotation: {
+      const InteropTypeBoundsAnnotation *Annot =
+        dyn_cast<InteropTypeBoundsAnnotation>(Bounds);
+      assert(Annot && "unexpected dyn_cast failure");
+      if (Annot != nullptr)
+        ResultType = Annot->getType();
+      break;
+    }
+    case BoundsExpr::Kind::ByteCount:
+    case BoundsExpr::Kind::ElementCount:
+    case BoundsExpr::Kind::Range: {
+      if (const PointerType  *PtrType = Ty->getAs<PointerType>()) {
+        if (PtrType->isUnchecked()) {
+          ResultType = Context.getPointerType(PtrType->getPointeeType(),
+                                              CheckedPointerKind::Array);
+          ResultType.setLocalFastQualifiers(Ty.getCVRQualifiers());
         }
-        break;
       }
-      default:
-        break;
+      else if (Ty->isConstantArrayType() || Ty->isIncompleteArrayType())
```

Comment:
Can we really do "interop-type inference" for incomplete arrays?

Reply:
Yes, for parameters.   For an array with variable length, you can write:
```
void f(int arr[], int len)
```
In Checked C this might become:
```
void f(int arr[] : count(len), int len)
```
By the way, it is considered bad form in C to write something like:
```
void f(int arr[5]);  // deceptive, no checking that callers actually
                     // pass arrays with 5 elements.
```
There's no guarantee in C that `arr` has 5 elements.   There's no language-level checking required and some C compilers will allow things like:
```
void bad_call(void) {
  int arr[4];
  f(arr);
}
```
Clang provides some limited checking , but it'll let you pass a pointer to f without any complaints
```
sketchy_call(int *p) {
  f(p);
}
```